apollo-router/src/apollo_telemetry.rs

```diff
@@ -0,0 +1,409 @@
+        // for more details and description.
+        let jh = tokio::task::spawn_blocking(|| {
+            opentelemetry::global::force_flush_tracer_provider();
+            opentelemetry::global::set_tracer_provider(provider);
+        });
+        futures::executor::block_on(jh)?;
+
+        Ok(tracer)
+    }
+
+    // XXX CANNOT USE SIMPLE WITH OUR IMPLEMENTATION AS NO RUNTIME EXISTS
+    // WHEN TRYING TO EXPORT...
+    /// Install the apollo telemetry exporter pipeline with the recommended defaults.
+    #[allow(dead_code)]
+    pub fn install_simple(mut self) -> Result<sdk::trace::Tracer, ApolloError> {
+        let exporter = self.get_exporter()?;
+
+        let mut provider_builder =
+            sdk::trace::TracerProvider::builder().with_simple_exporter(exporter);
+        if let Some(config) = self.trace_config.take() {
+            provider_builder = provider_builder.with_config(config);
+        }
+        let provider = provider_builder.build();
+
+        let tracer = provider.tracer("apollo-opentelemetry", Some(env!("CARGO_PKG_VERSION")));
+        let _prev_global_provider = global::set_tracer_provider(provider);
+
+        Ok(tracer)
+    }
+
+    /// Create a client to talk to our relay and return an exporter.
+    pub fn get_exporter(&self) -> Result<Exporter, ApolloError> {
+        let collector = match self.studio_config.clone() {
+            Some(cfg) => cfg.collector,
+            None => DEFAULT_SERVER_URL.to_string(),
+        };
+        let graph = self.graph_config.clone();
+
+        tracing::debug!("collector: {}", collector);
+        tracing::debug!("graph: {:?}", graph);
+
+        Ok(Exporter::new(collector, graph))
+    }
+}
+
+/// A [`SpanExporter`] that writes to [`Reporter`].
+///
+/// [`SpanExporter`]: super::SpanExporter
+/// [`Reporter`]: apollo_relay::Reporter
+#[derive(Debug)]
+pub struct Exporter {
+    collector: String,
+    graph: Option<StudioGraph>,
+}
+
+impl Exporter {
+    /// Create a new apollo telemetry `Exporter`.
+    pub fn new(collector: String, graph: Option<StudioGraph>) -> Self {
+        Self { collector, graph }
+    }
+}
+
+/// Apollo Telemetry exporter's error
+#[derive(thiserror::Error, Debug)]
+#[error(transparent)]
+pub struct ApolloError(#[from] apollo_relay::ReporterError);
+
+impl From<std::io::Error> for ApolloError {
+    fn from(error: std::io::Error) -> Self {
+        ApolloError(error.into())
+    }
+}
+
+impl From<JoinError> for ApolloError {
+    fn from(error: JoinError) -> Self {
+        ApolloError(error.into())
+    }
+}
+
+impl ExportError for ApolloError {
+    fn exporter_name(&self) -> &'static str {
+        "apollo-telemetry"
+    }
+}
+
+impl From<&StudioGraph> for ReporterGraph {
+    fn from(graph: &StudioGraph) -> Self {
+        ReporterGraph {
+            reference: graph.reference.clone(),
+            key: graph.key.clone(),
+        }
+    }
+}
+
+#[async_trait]
+impl SpanExporter for Exporter {
```

Comment:
I skipped most of this part for now because the end of the week, so I haven't reviewed any part of the code that actually builds up traces or stats.

Reply:
That code isn't correct or even full implemented yet. I'll need to dig further into that at some point. I'm fairly happy with the actual "transfer" data to our our ingress part now, so cleaning up the stats is my next objective.

Reply:
I've implemented the phase 1 work now, so I'll mark this has resolved.